### DISCUSSION TITLE
Cannot open REPL terminal on my Ubuntu server

### DISCUSSION THREAD

**Discussion by banishlight** [micropython/micropython#15504] [ESP32] [answered]
I have a ESP32-CAM connected with the ESP32-CAM-MB adapter.  On my macbook I can connect to my REPL terminal using
```screen /dev/tty.usbserial-142420 115200``` 
but on my Ubuntu server where I want to do my development I try
```screen /dev/ttyUSB0 115200```
and the terminal is just blank and I am not sure why...
I gave myself dialout permissions and I confirmed that it is connected using
```
sudo dmesg
```
I get the following:
```
[446863.355054] usb 3-7: new full-speed USB device number 4 using xhci_hcd
[446863.503894] usb 3-7: New USB device found, idVendor=1a86, idProduct=7523, bcdDevice= 2.64
[446863.503903] usb 3-7: New USB device strings: Mfr=0, Product=2, SerialNumber=0
[446863.503907] usb 3-7: Product: USB Serial
[446863.504638] ch341 3-7:1.0: ch341-uart converter detected
[446863.505114] usb 3-7: ch341-uart converter now attached to ttyUSB0
```
not sure why it is acting different on this machine.

**Comment by ricksorensen** [comment]
/dev/ttyACM0 instead of /dev/ttyUSB0?

**Reply by davefes** [reply]
Had a similar problem and Google led me to modemmanager.  I purged that and problem seemed to go away.

**Reply by banishlight** [reply]
Don't have any ACM0 device..
Also did this to remove modemmanager and still no success:
```
sudo systemctl stop ModemManager
sudo systemctl disable ModemManager
```

**Reply by davefes** [reply]
I don't think that removes it.  Try purge.

**Reply by banishlight** [reply]
just ran
```
sudo apt-get purge modemmanager
```
no changes so far.

**Reply by davefes** [reply]
Which version of Ubuntu?

**Reply by banishlight** [reply]
Ubuntu 22.04

**Reply by davefes** [reply]
I remember now there was another step"
https://github.com/orgs/micropython/discussions/13418
`brltty`

**Reply by banishlight** [reply]
Never got any message in dmesg about br1tty so I never touched it so far.
Tried to remove it now and got this:
```
$ sudo apt remove brltty
Package 'brltty' is not installed, so not removed
```

**Reply by davefes** [reply]
OK, I tried your command on an ESP32-CAM and likewise a blank screen.

I use rshell and had to do the previously mentioned steps.

**Reply by davefes** [reply]
I used your command on ESP32 Generic and got a blank screen AND then did a hard reset and got REPL.  But couldn't see the filesystem.

Good luck.

**Reply by banishlight** [reply]
So I should look at getting rshell setup on my Ubuntu server?

**Reply by davefes** [reply]
`mpremote` is the official package.  I have been using `rshell` for a couple of years.  I tried `mpremote` but reverted!

Probably more appropriate to go with `mpremote' as you are starting fresh.

If you do go with `rshell` in .bashrc add:
`alias rsusb='rshell -p /dev/ttyUSB0 --baud=115200 --editor nano --buffer-size=30 -a'
`
and then:
`source ~/.bashrc`

**Reply by banishlight** [reply]
```
Connecting to /dev/ttyUSB0 (buffer-size 30)...
Trying to connect to REPL ....................Unable to connect to REPL
Welcome to rshell. Use Control-D (or the exit command) to exit rshell.

No MicroPython boards connected - use the connect command to add one
```
Any idea what is going on?

**Reply by davefes** [reply]
rsusb
enter
and within a few seconds press the EN button

**Reply by banishlight** [reply]
Also what kind of hard reset did you do?  The reset button on the ESP32-CAM-MB adapter?

**Reply by davefes** [reply]
Yes.

**Reply by davefes** [reply]
Hard reset is pressing the button by the 3V3 regulator.

**Reply by Josverl** [reply]
On ubuntu desktop I have the experience that I do need to remove the braille reader support as that driver has rather broad assumptions about what a braille reader is. It assumes that anything connected using a `CP210x USB to UART Bridge` is considered to be a braille reader , and it then locks that device.
The CP210X is often combined with a esp8266, but not exclusively, I 've seen many different combinations, including a few with an ESP32 IIRC.

While I complete agree that accessibility is part of an OS, it appears to be a rather assuming implementation.
note that I'm not sure if `brltty` is part of the server image as well, if not apt will let you know.

My documented fix is: 
Remove conflicting packages
```
# remove braille support as it conflicts with esp8266
sudo apt remove brltty
```

**Reply by banishlight** [reply]
- With ```screen /dev/ttyUSB0 115200``` && reset button
- ```rsusb```
no connection, blank display on my terminal...

**Reply by davefes** [reply]
When you do a `rsusb` do you get any messages?

Which camera firmware did you put on the ESP32-CAM and you removed the link for flashing.

**Reply by banishlight** [reply]
This is what I get from rsusb:
```
$ rsusb
Using buffer-size of 30
Connecting to /dev/ttyUSB0 (buffer-size 30)...
Trying to connect to REPL ....................Unable to connect to REPL
Welcome to rshell. Use Control-D (or the exit command) to exit rshell.

No MicroPython boards connected - use the connect command to add one
```
and I am not sure how to remove the link for flashing with the board that I am using to connect via usb

**Comment by shariltumin** [comment, marked as answer]
ESP32-CAM with ESP32-CAM-MB adapter will not give a REPL because the Linux serial driver sets RTS and DTR to high by default. This means that the board is in programming mode when connected to USB. This makes it easy to flash new firmware, but does not reset the board to normal operating mode.

This is what you can do (assuming the ESP32-CAM-MB is connected to /dev/ttyUSB0):

```bash
$ miniterm --rts 0 --dtr 0 /dev/ttyUSB0 115200
```

You are still in boot loader mode. We need to do a hard reset. Locate the reset button on the ESP32-CAM (underneath the PCB, marked RST, near the 5V PIN). Press this button to initiate a hard reset. The RST on the ESP32-CAM-MB does not work.

"Ctrl+alt+)" (which is Ctrl+]) to quit miniterm. Now

```bash
$ screen /dev/ttyUSB0 115200
```
will give you REPL after a few 'Enter' keystrokes, "Ctrl-a k" will quit screen.

Now all serial terminal programs; tio, mpremote, Thonny etc. will work normally.

Note that you can also try ```gtkterm -p /dev/ttyUSB0``` and unset DTR and RTS from "Control signals" menu.

Hope this helps.

**Reply by banishlight** [reply]
You are absolutely right this actually got me into my REPL 
```miniterm --rts 0 --dtr 0 /dev/ttyUSB0 115200```
incredible!
I still cannot connect using ```screen /dev/ttyUSB0 115200``` but I will post what I find from here thank you :)
